Updated metadata documentation part 1 - introduction

bjwheltor opened the pull request:
Description

An initial draft of the Introduction metadata document (a re-working of information currently hosted on an internal Met Office site). This creates a new `metadata` folder under `doc/source`, which contains the document `Introduction.rst` and a number of text files holding modified `ncdump` and iris `print(cube)` output from which snippets are pulled into the documentation.

bayliffe commented on doc/source/metadata/Introduction.rst:
```diff
@@ -0,0 +1,724 @@
+wmo_id
+    For WMO sites, this is the numerical WMO identifier stored as a string; 
+    for other sites this is a set to ``NaN`` (at present, the formatting of this is not ideal, but will be changed to ensure that the numbers are right-justified and padded with zeros in the next release). 
+
```
Needs a section on the `met_office_site_id`. It is worth highlighting here that the `met_office_site_id` name is user configurable, such that it can be changed for different institutions / indices. It is a zero-padded string.

bjwheltor replied:
Added. Thanks.

bayliffe commented on doc/source/metadata/Introduction.rst:
```diff
@@ -0,0 +1,724 @@
+    although this is not explicit in the metadata. 
+
+wmo_id
+    For WMO sites, this is the numerical WMO identifier stored as a string; 
+    for other sites this is a set to ``NaN`` (at present, the formatting of this is not ideal, but will be changed to ensure that the numbers are right-justified and padded with zeros in the next release). 
```
I assume this is just out of date? These are 5-character strings, zero-padded ID numbers. The non WMO sites are set to the string "None" as iris cannot save NaN.

bjwheltor replied:
Updated. Thanks.

bayliffe commented on doc/source/metadata/Introduction.rst:
```diff
@@ -0,0 +1,724 @@
+
+
+The global attribute ``mosg__model_configuration`` is always a single model 
+(for the Met Office ``uk_ens`` or ``gl_ens`` 
```
```suggestion
(for the Met Office ``uk_ens`` or ``gl_ens``)
```

bayliffe commented on doc/source/metadata/Introduction.rst:
```diff
@@ -0,0 +1,724 @@
+
+For gridded data only, the ``title`` only identifies the single model used
+(for the Met Office, this would be either
+``MOGREPS-UK Model Forecast on UK 2 km Standard Grid`` or 
+``MOGREPS-G Model Forecast on UK 2 km Standard Grid``),
```
These are out of date. The titles now include the `Post-processed` prefix to avoid issues with visual weather distinguishing between the raw model output and the IMPROVER products. We have done processing here, so the title is not unreasonably applied.

bjwheltor replied:
Thanks. It's in my sample data, but forgot about the change.

bayliffe commented on doc/source/metadata/Introduction.rst:
```diff
@@ -0,0 +1,724 @@
+The main variable has a ``cell_methods`` attribute, set to
+``realization: mean``.
+
+At present, there is no ``blend_time`` coordinate variable, 
+so, st present, an idea of the time at which the data was generated 
+can be obtained from the ``forecast_reference_time``
```
```suggestion
At present there is no ``blend_time`` coordinate variable.
For now an idea of the time at which the data was generated 
can be obtained from the ``forecast_reference_time``
```

bayliffe commented on doc/source/metadata/temp12max_perc_ncdump.txt:
```diff
@@ -0,0 +1,82 @@
+netcdf PERC-20221222T1800Z-B20221220T0615Z-temperature_at_screen_level_max-PT12H {
```
I think your file has a `PERC-` prefix which shows up weirdly here. Not a major issue.

bjwheltor replied:
I've removed these prefixes from the sample code now.

bayliffe reviewed (changes requested):
Bruce, I've added some detail about where MO specific stuff can be adapted by other users. I've asked for a few changes, and added a load of typo corrections.

I'm going to rebuild the `improver_production` environment that we all use such that the local compilation of the docs works with bulleted lists etc., more inline with how it appears online.

bayliffe commented:
@bjwheltor you also need to add your name to the [contributors list](https://github.com/bjwheltor/improver/blob/metadata_doc_rev2/CONTRIBUTING.md) as this is causing the tests to fail at present.

bjwheltor commented on doc/source/metadata/Introduction.rst:
```diff
@@ -0,0 +1,724 @@
+
+time
+    The time at which the parameter value is valid.
+    As IMPROVER may blend data with different reference times 
+    (used at the Met Office to generate a continually updating dataset,
+    with a single value for each time),
+    this is really the only piece of time information that matters
+    for most use cases. 
```
Fair enough

bjwheltor commented on doc/source/metadata/Introduction.rst:
```diff
@@ -0,0 +1,724 @@
+Note that to aid readability, in both cases
+``probability_of_air_temperature_above_threshold`` 
+has been shortened to ``prob_temp12m``
+and for the ncdump output trailing semicolons have also been removed
```
This was mainly done to fit this on a page - originally this was in a Word document. I maintained it to avoid scroll bars on the code snippets, but agree that a more pure approach is to show people what they would actually see when invoking an ncdump command, so I'll revert these to pure copies of the ncdump files (including semicolons, full titles and without artificial formatting or line breaks).

bjwheltor commented on doc/source/metadata/Introduction.rst:
```diff
@@ -0,0 +1,724 @@
+has been shortened to ``prob_temp12m``
+and for the ncdump output trailing semicolons have also been removed
+and the text for the global attribute
+``mosg__model_run`` has line breaks inserted.
```
I've removed the line breaks, as they are artificial. It means that you get a single line of text that goes off the viewing area, but it's what you'd actually see and being at the bottom of the file, the scroll bar is close at hand!

bjwheltor commented on doc/source/metadata/Introduction.rst:
```diff
@@ -0,0 +1,724 @@
+.. literalinclude:: temp12max_prob_ncdump.txt
+    :tab-width: 4
+    :lines: 66-79
+    :emphasize-lines: 2, 12-14
```
I agree, that's much better. I've changed it to "friendly"

bjwheltor commented on doc/source/metadata/Introduction.rst:
```diff
@@ -0,0 +1,724 @@
+    A specific model is specified where data is from a single model 
+    and no significant post-processing has been applied. 
+
+The other two attributes are specific to IMPROVER,
+originally used in separate Met Office process to 'standardise'
+the model output, which is why they are prefixed by ``mosg__``. 
```
1. I do mean StaGE, and I added this in to try to explain the rather standard structure of the attribute names, but if it's confusing then it's better removed. Done.
2. I've included a paraphrased version of what you've said below the attribute description.

bjwheltor commented on doc/source/metadata/Introduction.rst:
```diff
@@ -0,0 +1,724 @@
+    Number of points in the horizontal x-direction
+
+projection_x_coordinate
+    Number of points in the horizontal y-direction. 
```
I think it's a bit more subtle than this. Coordinate variables can be flagged as being horizontal with axis = "X"/"Y" (which we do) and vertical with  axis = "Z" (which we don't do), but can be called anything. These are usually given sensible variable names and the dimensions should have the same name as the coordinate variable. There are standard names for the usual coordinate variables, However, taken together, I'm not convinced that the Convention stops using confusing names if you want to! 

I may have missed your point here, so feel free to come back with specific suggestion. However, I suspect any addition should go in the coordinate variables section rather then the dimensions section.

bjwheltor commented on doc/source/metadata/Introduction.rst:
```diff
@@ -0,0 +1,724 @@
+
+units
+    the units of measurement for the quantity.
+    These will always be SI units. 
```
Yes. Added.

bjwheltor commented on doc/source/metadata/Introduction.rst:
```diff
@@ -0,0 +1,724 @@
+
+In summary these are:
+
+* threshold 
+* projection_y_coordinate 
+* projection_x_coordinate 
+* height 
+* time 
+* blend_time 
+* forecast_period 
+* forecast_reference_time 
```
Fair point. I think I was presenting them in an order that I considered (rather arbitrarily) to be logical (and pushing the deprecated ones to the end), but I agree it makes more sense to stick with the order in the example, so I've changed it.

bjwheltor commented on doc/source/metadata/Introduction.rst:
```diff
@@ -0,0 +1,724 @@
+a further dimensional coordinate variable is required to allow the data 
+to be fully interpreted.
+This holds the set of thresholds for the probabilities,
+which in this case are a set of 1-hour precipition accumulation values.
+Both, these are shown in the code snippets below
```
Corrected.

bjwheltor commented on doc/source/metadata/Introduction.rst:
```diff
@@ -0,0 +1,724 @@
+    This usually represent the interval between the ``forecast_reference_time``
+    and the validity time (``time``), but as stated above,
+    there is no unique ``forecast_reference_time``, and forecasts valid at
+    different times may have a different ``blend_time``, so at brackets
```
Should say "at best"; corrected.

bjwheltor commented on doc/source/metadata/Introduction.rst:
```diff
@@ -0,0 +1,724 @@
+(but would be latitude and longitude for an equi-rectangular 
+projection grid, such as the Met Offuce global domain),
+provide the coordinates of the grid points at the centre of the grid cell,
+with teo further variables defining the cell bounds.
```
corrected erroneous spelling of "two", as well as interesting spelling of "Office"!

bjwheltor commented on doc/source/metadata/Introduction.rst:
```diff
@@ -0,0 +1,724 @@
+    A descriptive name, either from the `CF Standard Name`_ list.
+
+units
+    The units of measure for the quantity, these will always be SI units. 
```
Added text to this effect.

bjwheltor commented on doc/source/metadata/Introduction.rst:
```diff
@@ -0,0 +1,724 @@
+    The units of measure for the quantity, these will always be SI units.  
+
+positive
+    Indicates the direction in which values of the vertical coordinate increase. 
```
Added in your example

bjwheltor commented on doc/source/metadata/Introduction.rst:
```diff
@@ -0,0 +1,724 @@
+    holding WMO identifier and Met Office identifiers, respectively.
+
+.. how tied to the code are these identifiers, as they may not be appropiate
+    for other users?
```
Added in the later section on this attribute.

bjwheltor commented:
> @bjwheltor you also need to add your name to the [contributors list](https://github.com/bjwheltor/improver/blob/metadata_doc_rev2/CONTRIBUTING.md) as this is causing the tests to fail at present.

Done.

bjwheltor commented:
@bayliffe - Thanks for all the useful comments (and sorry about all the typos). I believe that I have now addressed all your review comments, apart from the one regarding x/y coordinate variables (which I either disagree with, or I missed the point - see comment for specifics). Hopefully, the PR is in reasonable shape, but I did manage to get the local and remote versions out of step by apply changes both through Github and Git (now resolved) - I'll know better next time. Let me know if you spot any immediate errors, but I expect the fresh eyes of a second reviewer are more likely to discover further issues!

bayliffe commented on doc/source/metadata/Introduction.rst:
```diff
@@ -0,0 +1,735 @@
+(although the spot data is still labelled as IMPROVER,
+as a site extraction has taken place). 
+
+As this is mean, there is no dimension and coordinate variable for 
```
```suggestion
As this is a mean, there is no dimension and coordinate variable for 
```

bjwheltor replied:
Much as I wanted to leave this as "this is mean", I have changed it to the "a" in.

bayliffe reviewed (changes requested):
A few typos, but otherwise happy Bruce.  Thanks.

bjwheltor commented on doc/source/metadata/Introduction.rst:
```diff
@@ -0,0 +1,735 @@
+provides a fuller view
+(although iris can be used to fully explore the metadata).
+
+Note that to aid readability, in both cases
+``probability_of_air_temperature_above_threshold`` 
+has been shortened to ``prob_temp12m``
+and for the ncdump output trailing semicolons have also been removed
+and the text for the global attribute
+``mosg__model_run`` has line breaks inserted.
```
Removed section

bjwheltor commented on doc/source/metadata/Introduction.rst:
```diff
@@ -0,0 +1,735 @@
+    holding zero padded WMO identifier and Met Office identifiers, respectively.
+
+.. how tied to the code are these identifiers, as they may not be appropiate
+    for other users?
```
Done

bjwheltor commented on doc/source/metadata/Introduction.rst:
```diff
@@ -0,0 +1,735 @@
+a further dimensional coordinate variable is required to allow the data 
+to be fully interpreted.
+This holds the set of thresholds for the probabilities,
+which in this case are a set of 12-hour maxiumum temperature values.
```
Corrected.

bjwheltor commented on doc/source/metadata/Introduction.rst:
```diff
@@ -0,0 +1,735 @@
+    the ``blend_time`` will not be the same for all validity times,
+    as forecasts in the near future are updated more frequently).
+    Strictly, as ``blend_time`` is not part of the `CF Metadata Conventions`_
+    it should arguable include a 'namespace element' such as ``mosg__``
```
Corrected.

bjwheltor commented on doc/source/metadata/Introduction.rst:
```diff
@@ -0,0 +1,735 @@
+
+Two horizontal coordinate variables, 
+here projection_y_coordinate and projection_x_coordinate
+(but would be latitude and longitude for an equi-rectangular 
```
I've changed both instances to "equirectangular" as most sources don't capitalise the name. Nor indeed do they for the other projection names, with the exception of where they include the name of a person, so I've also updated the other projection reference to "Lambert azimuthal equal area".

bjwheltor commented on doc/source/metadata/Introduction.rst:
```diff
@@ -0,0 +1,735 @@
+Main percentile variable 
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+In the example here, the main variable is a 12 hour maxiumum temperature,
```
Corrected

bjwheltor commented on doc/source/metadata/Introduction.rst:
```diff
@@ -0,0 +1,735 @@
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+In the example here, the main variable is a 12 hour maxiumum temperature,
+but representing the nth percentile of the temperature forecast probability distribution
```
Changed.

bjwheltor commented on doc/source/metadata/Introduction.rst:
```diff
@@ -0,0 +1,735 @@
+
+Broadly, the spot data parallels the form of the gridded data, 
+both probabilities and percentiles are produced,
+but there are some notable differences because of the nature of data.
```
added in "the"

bjwheltor commented:
@bayliffe  - hopefully, I've now addressed the latest set of changes and updated the PR. Thanks.

bayliffe commented on doc/source/metadata/Introduction.rst:
```diff
@@ -491,13 +483,13 @@ percentile
 .. literalinclude:: temp12max_perc_ncdump.txt
     :tab-width: 4
-    :lines: 8-14
+    :lines: 8-14s
```
I think this addition of a trailing `s` was a mistake.

bjwheltor replied:
Indeed - probably CNTL S without the CNTL. Now corrected.

brhooper commented on doc/source/metadata/Introduction.rst:
```diff
@@ -0,0 +1,723 @@
+    Used to describe statistical processing applied to the quantity
+    no longer requires the additional non-standardized part, 
+    ``(comment: of air_temperature)``, 
+    as they now refer to the main variable that is a maximum temperature over 12 hours. 
```
I think this needs re-wording. Would the following suggestion work?

```suggestion
Used to describe statistical post-processing applied to the quantity. Cell methods no longer require the additional non-standardized part, ''(comment: of air_temperature)'', as they now refer to the main variable that is a maximum temperature over 12 hours.
```

brhooper commented on doc/source/metadata/Introduction.rst:
```diff
@@ -0,0 +1,723 @@
+
+bounds
+    Just a pointer to the variable defining the start and end of the time period,
+    if it is present. 
```
I think that this could be shortened by writing:
```suggestion
   Pointer to the variable defining the start and end of the time period, if it is present. 
```
however this is only personal preference so I am happy if you prefer to leave it as is. If it is changed, there are a few other instances of sentences beginning "Just a..." which should also be changed.

bjwheltor replied:
Done - in fact I shortened it slightly further.

brhooper commented on doc/source/metadata/Introduction.rst:
```diff
@@ -0,0 +1,723 @@
+    Cartesian coordinate.  
+
+bounds
+    Just a pointer to the variable defining the edges of the grid cells 
```
See previous comment on sentences starting with "Just a..."

bjwheltor replied:
Also changed.

brhooper commented on doc/source/metadata/Introduction.rst:
```diff
@@ -0,0 +1,723 @@
+    The dimension for the index for the set of sites. 
+
+string5 / string8
+    Just constants used to dimension the character length of the string variable
```
See previous comments on starting sentence with "Just..."

bjwheltor replied:
And also changed.

brhooper commented on doc/source/metadata/Introduction.rst:
```diff
@@ -0,0 +1,723 @@
+    exists as a descriptive name in the `CF Standard Name`_ list. 
+
+cell_methods
+    Used to describe statistical processing applied to the quantity
```
I think this needs re-wording. Would the following suggestion work?

```suggestion
Used to describe statistical post-processing applied to the quantity. Cell methods no longer require the additional non-standardized part, ''(comment: of air_temperature)'', as they now refer to the main variable that is a maximum temperature over 12 hours.
```

bjwheltor replied:
I think I've updated this as suggested.

brhooper reviewed (changes requested):
One part which I think requires slightly re-wording. I also highlighted a few things which could be changed but I think these are more down to my personal preference so I am also happy if you prefer to leave as is.

bjwheltor commented:
@brhooper - I think that I've now addressed your review comments and committed change to the PR, so if you're happy with these, please resolve and merge.

brhooper reviewed (approved):
Thanks for making those changes - I'm happy with this now
